Added static registration for http tracers. (#967)

mrice32 opened the pull request:
Added a static registration mechanism (following the pattern of filters mentioned in the comments on #967) for http tracers.  This will likely be expanded to a more general registration mechanism in future work.  This PR is part of the work towards solving #967.

mattklein123 commented:
I think @htuch will take first pass, but meta comment, if this goes in I think https://github.com/lyft/envoy/pull/982 can lose the defines and just not compile lightstep code. I think this will also solve @lizan issue.

jamessynge commented on source/server/config/http/lightstep_http_tracer.cc:
```diff
@@ -0,0 +1,43 @@
+
+  Envoy::Runtime::RandomGenerator& rand = server.random();
+
+  std::unique_ptr<lightstep::TracerOptions> opts(new lightstep::TracerOptions());
```
Please include lightstep/options.h and lightstep/tracer.h

jamessynge commented on source/server/configuration_impl.h:
```diff
@@ -44,6 +45,19 @@ class NetworkFilterConfigFactory {
+ * Implemented by each network filter and registered via registerHttpTracerFactory() or
+ * the convenience class RegisterHttpTracerFactory.
+ */
+class HttpTracerFactory {
```
Is there a good reason to NOT put this in include/envoy/tracing/http_tracer.h?

mrice32 replied:
It creates a build dependency cycle because //include/envoy/tracing:http_tracer_interface and //include/envoy/server:instance_interface would both depend on one another.

jamessynge commented on source/server/configuration_impl.h:
```diff
@@ -44,6 +45,19 @@ class NetworkFilterConfigFactory {
 };
 
 /**
+ * Implemented by each network filter and registered via registerHttpTracerFactory() or
```
Implemented for each kind of HttpTracer.

jamessynge commented on source/server/configuration_impl.h:
```diff
@@ -44,6 +45,19 @@ class NetworkFilterConfigFactory {
+public:
+  virtual ~HttpTracerFactory() {}
+
+  virtual Tracing::HttpTracerPtr
```
Please document the method (expected behavior, error conditions).

jamessynge commented on source/server/configuration_impl.cc:
```diff
@@ -116,32 +116,20 @@ void MainImpl::initializeTracers(const Json::Object& configuration) {
+    if (http_tracer) {
+      http_tracer_ = std::move(http_tracer);
+      found_tracer = true;
+    }
```
break;

mattklein123 commented on source/server/configuration_impl.h:
```diff
@@ -135,6 +159,11 @@ class MainImpl : Logger::Loggable<Logger::Id::config>, public Main {
   }
 
+  static std::list<HttpTracerFactory*>& httpTracerFactories() {
+    static std::list<HttpTracerFactory*> http_tracer_factories;
```
If we are following the letter of the law here I think this should be a pointer and the list allocated with new

mattklein123 commented:
@mrice32 also please sign the CLA

htuch commented on source/server/configuration_impl.h:
```diff
@@ -162,6 +191,17 @@ template <class T> class RegisterNetworkFilterConfigFactory {
+template <class T> class RegisterHttpTracerFactory {
+public:
+  RegisterHttpTracerFactory() {
+    static T instance;
```
In addition to the above `static std::list`, this should also be created on use. For reference:

https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables
https://isocpp.org/wiki/faq/ctors#static-init-order-on-first-use

htuch commented on source/server/config/http/zipkin_http_tracer.h:
```diff
@@ -0,0 +1,27 @@
+class ZipkinHttpTracerFactory : public HttpTracerFactory {
+public:
+  // HttpTracerFactory
+  Tracing::HttpTracerPtr tryCreateHttpTracer(const std::string& type,
```
I'm a bit surprised by the interface (although it's consistent with the existing factory interface for filters). How come we need to probe each individual filter/tracer to figure out if they match type, rather than registering a map from type to filter/tracer? @mattklein123 for the filter interface.

mattklein123 replied:
We should do that. :) I have no idea why I did it this way in the first place. Very long time ago now. Feel free to fix in this PR or a follow up.

mrice32 replied:
Okay, cool.  I'll add it into this PR.

mattklein123 replied:
This has the added benefit of you can fail if you try to register something with the same name. So would be great to add test for that also, and fix all the networking filters too. Bigger change, but I think well worth it.

mrice32 commented:
Took a stab at the map static registration mechanism, PTAL.

ccaraman commented:
CLA has been signed.

RomanDzhabarov commented on test/server/config/network/config_test.cc:
```diff
@@ -323,11 +319,14 @@ TEST(NetworkFilterConfigTest, BadAccessLogNestedTypes) {
 }
 
+TEST(NetworkFilterConfigTest, DoubleRegistrationTest) {
+  EXPECT_THROW(RegisterNetworkFilterConfigFactory<ClientSslAuthConfigFactory>(), EnvoyException);
```
nit: I think EXPECT_THROW_WITH_MESSAGE makes a test slightly better (that macro is from common_test/utility.h)
In Envoy we throw EnvoyExceptions for various reasons it's better to verify that it's what we expect.

RomanDzhabarov commented on source/server/configuration_impl.h:
```diff
@@ -37,10 +40,42 @@ class NetworkFilterConfigFactory {
-                                                        const Json::Object& config,
-                                                        Server::Instance& server) PURE;
+  /**
+  * Create a particular network filter factory implementation.  If the creation fails, an error will
```
can you be more specific, is it only EnvoyException which can be thrown?

RomanDzhabarov commented on source/server/configuration_impl.h:
```diff
@@ -37,10 +40,42 @@ class NetworkFilterConfigFactory {
+  * Create a particular network filter factory implementation.  If the creation fails, an error will
+  * be thrown.  The returned callback should always be valid.
+  */
+  virtual NetworkFilterFactoryCb createFilterFactory(NetworkFilterType type,
```
i think we are a bit more strict now around documenting @params etc, can you add those

RomanDzhabarov commented on source/server/configuration_impl.h:
```diff
@@ -130,9 +185,16 @@ class MainImpl : Logger::Loggable<Logger::Id::config>, public Main {
-    static std::list<NetworkFilterConfigFactory*> filter_config_factories;
-    return filter_config_factories;
+  static std::unordered_map<std::string, NetworkFilterConfigFactory*>& filterConfigFactories() {
+    static std::unordered_map<std::string, NetworkFilterConfigFactory*>* filter_config_factories =
```
any cons that to be: static std::unordered_map<std::string, NetworkFilterConfigFactory*> filter_config_factories; ?

htuch replied:
Yeah, the static initialization order fiasco :) See https://isocpp.org/wiki/faq/ctors#static-init-order-on-first-use. We have a section on this now in https://github.com/lyft/envoy/blob/master/STYLE.md.

RomanDzhabarov replied:
oki doki, i have not inspected all the code, but looks like https://github.com/lyft/envoy/blob/master/source/server/config/network/http_connection_manager.h#L117 has a similar issue as you pointed out @htuch

RomanDzhabarov commented on source/server/config/network/http_connection_manager.cc:
```diff
@@ -152,21 +154,11 @@ HttpConnectionManagerConfig::HttpConnectionManagerConfig(const Json::Object& con
-      throw EnvoyException(fmt::format("unable to create http filter factory for '{}'/'{}'",
-                                       string_name, string_type));
-    }
+    HttpFilterConfigFactory* factory = filterConfigFactories().at(string_name);
```
do we have a test when we call it without "string_name" filtered registered

RomanDzhabarov commented on source/server/config/network/http_connection_manager.h:
```diff
@@ -48,10 +49,19 @@ class HttpFilterConfigFactory {
+  * Provides the identifying name for a particular implementation of an http filter produced by the
+  * factory.
+  */
+  virtual std::string name() PURE;
```
nbd (as there are calls to it only on a startup)
but this could be `const std::string&`

mrice32 replied:
I was curious about this.  We would need to define a string as a member of the class or make a static string, right?  If I'm thinking about it correctly, both of those options would ultimately be more copying than the current version (because we should only call the method once during registration and the string produced will be moved into the unordered_map as opposed to the copy required for a const std::string &).  However, if this is just a readability/API preference for envoy, that totally makes sense.

RomanDzhabarov replied:
yeah, we could do static const string in a class.
Thinking about this more, as it's called once there is no point of having ref returned, agreed.

htuch commented on source/server/config/http/dynamo.cc:
```diff
@@ -8,13 +8,13 @@ namespace Envoy {
+                                                            const Json::Object&,
+                                                            const std::string& stat_prefix,
+                                                            Server::Instance& server) {
+  if (type != HttpFilterType::Both) {
```
Unrelated to the review, a question that came up recently while discussing filters with @alyssawilk. Why is Envoy filter config structured such that (1) filters know their `HttpFilterType` yet (2) the type is specified in the config as well, meaning we need to check consistency. It seems this could be elided from the config. The only thing I can think of is that some filters might work as encoder/decoder depending upon configuration... @RomanDzhabarov @ccaraman @junr03 for insight here.

RomanDzhabarov replied:
@htuch cannot think about anything else, but fwiw, currently, no HTTP filters support this behavior.

mrice32 commented:
Thanks @RomanDzhabarov for the thorough review.  It's really helpful as I'm trying to learn the how things should be done in envoy.  I believe I have corrected for most of the comments except for one that I had a question about.  @htuch I have updated to add the backwards compatibility.  I documented the list registration related methods/classes as DEPRECATED in the comments.  Please let me know if there's some other place I should document the deprecation.  I will add a dummy test to ensure that the deprecated filter registration continues to work and commit it today.

RomanDzhabarov commented:
LGTM, @htuch for final pass
